Bluetooth: GATT: Document some macro arguments

spbnick opened the pull request:
Name some types in bluetooth/gatt.h and use those names to document the expected values of some of the BT_GATT_* macro arguments, in an attempt to make understanding the API a little easier.

BTW, could BT_GATT_CHRC_* macros be enum members instead?

Thalley commented on include/bluetooth/gatt.h:
```diff
@@ -100,44 +100,55 @@ enum {
+ *  @return Number of bytes read, or in case of an error
+ *          BT_GATT_ERR() with a specific ATT error code.
+ */
+typedef ssize_t (*bt_gatt_attr_read_func_t)(struct bt_conn *conn,
+					    const struct bt_gatt_attr *attr,
+					    void *buf, uint16_t len,
+					    uint16_t offset);
```
I like the changes you did for the enum and the macros, but I actually prefer defining callbacks as functions inside the scope of the callback structure, but we'll see if others agree/disagrees with this.

spbnick replied:
Thanks for the review, @Thalley! Yeah, I see how defining callbacks in place can be more compact. However, that way it's harder to refer to them from other places in the documentation. Let's see what others have to say.

BTW, while I (hopefully) have your ear, I'm struggling regenerating the Doxygen HTML locally after changing the header. I get `Doxygen build will be skipped (no changes)!` when running `make` in the `doc` directory (and I made sure there were changes). Even if I edit the code detecting the changes and doxygen runs, the results apparently are not copied to the output directory anyway. What am I doing wrong? Are there any docs on this I should read?

Resolving this could help me fix the issues CI has found.
Thanks!

Thalley replied:
@spbnick Agreed, it is harder to refer to a field in a struct, but you can do something like `@ref <name_of_struct>::<name_of_field>` 
So instead of `Descriptor attribute write callback (bt_gatt_attr_write_func_t)` you could do `@ref bt_gatt_attr::write` I think. 

I haven't built documentation locally myself for a very long time, so not sure if I can help you with that issue

spbnick replied:
> @spbnick Agreed, it is harder to refer to a field in a struct, but you can do something like @ref <name_of_struct>::<name_of_field>
Ah, I thought there should be something like that. I always forget the Doxygen refernce syntax. Yeah, I could do it like that, if there's preference. There's one more benefit for having the type defined separately, though: you can then define your own variables, e.g. for your own structures or functions generating those attributes.

alwa-nordic replied:
I think you have to use the dot separator in Doxygen i.e. `@ref type_name.field_name`.

spbnick commented:
Alright, I seem to have fixed the Doxygen issue, but only through the process of elimination, not reading (or even grepping) Doxygen docs. I would be grateful if somebody could enlighten me why the following change breaks Doxygen build:
```patch
diff --git a/include/bluetooth/gatt.h b/include/bluetooth/gatt.h
index 9d8e223433..61b2d3b57b 100644
--- a/include/bluetooth/gatt.h
+++ b/include/bluetooth/gatt.h
@@ -671,7 +671,7 @@ ssize_t bt_gatt_attr_read_chrc(struct bt_conn *conn,
  *
  *  @param _uuid Characteristic attribute uuid.
  *  @param _props Characteristic attribute properties,
- *               a bitmap of BT_GATT_CHRC* macros.
+ *               a bitmap of BT_GATT_CHRC_* macros.
  *  @param _perm Characteristic Attribute access permissions,
  *              a bitmap of enum bt_gatt_perm values.
  *  @param _read Characteristic Attribute read callback
```
Thank you!

nashif commented:
> . I would be grateful if somebody could enlighten me why the following change breaks Doxygen build:

@gmarull  any idea?

gmarull commented:
Can you try `@c BT_GATT_CHRC_*` (make it a literal)?

Thalley commented on include/bluetooth/gatt.h:
```diff
@@ -100,44 +100,55 @@ enum {
+ *  @param offset Offset to start reading from
+ *
+ *  @return Number of bytes read, or in case of an error
+ *          BT_GATT_ERR() with a specific ATT error code.
```
```suggestion
 *          BT_GATT_ERR() with a specific BT_ATT_ERR_* error code.
```

Thalley commented on include/bluetooth/gatt.h:
```diff
@@ -100,44 +100,55 @@ enum {
+ *  @param flags  Flags (BT_GATT_WRITE_*)
+ *
+ *  @return Number of bytes written, or in case of an error
+ *          BT_GATT_ERR() with a specific ATT error code.
```
```suggestion
 *          BT_GATT_ERR() with a specific BT_ATT_ERR_* error code.
```

asbjornsabo commented on include/bluetooth/gatt.h:
```diff
@@ -100,44 +100,55 @@ enum {
-			 const void *buf, uint16_t len, uint16_t offset,
-			 uint8_t flags);
-
+	/** Attribute read callback */
```
Personally, I am happy to have the callback function definitions within the struct, so I do not see it as necessary to move them out.

spbnick replied:
My idea was to allow easier reuse of the function (pointer) types without having to copy the definitions to the user code, plus easier reference from other parts of Doxygen docs. You would be the second person preferring them inside the struct, so if these arguments don't sway you, I'll leave them in place :slightly_smiling_face:

asbjornsabo reviewed (commented):
Improving the documentation of GATT is a Good Thing(TM) :-)

spbnick commented:
Thank you, @gmarull.

> Can you try `@c BT_GATT_CHRC_*` (make it a literal)?

[This helped](https://github.com/zephyrproject-rtos/zephyr/pull/40436/commits/4fe3074c62129c754aad6ff7e89de336981f42f5) (docs built), and I'll use it, but I would still like to know what was going on :joy:.

spbnick committed:
Bluetooth: GATT: Spell out BT_ATT_ERR_*

Spell out BT_ATT_ERR_* instead of using less specific "ATT error code".

Co-authored-by: Emil Gydesen <Thalley@users.noreply.github.com>
Signed-off-by: Nikolai Kondrashov <spbnick@gmail.com>

Thalley commented on include/bluetooth/gatt.h:
```diff
@@ -108,44 +108,55 @@ enum {
+ *  @param buf    Buffer with the data to write
+ *  @param len    Number of bytes in the buffer
+ *  @param offset Offset to start writing from
+ *  @param flags  Flags (BT_GATT_WRITE_*)
```
I see that `BT_GATT_WRITE_` is used to refer to the flags in multiple places, but we could use `BT_GATT_WRITE_FLAG_` to be more specific, and avoid any issues if any other enum with `BT_GATT_WRITE_` would be introduced.

Thalley commented on include/bluetooth/gatt.h:
```diff
@@ -925,9 +943,12 @@ ssize_t bt_gatt_attr_read_cpf(struct bt_conn *conn,
+ *  @param _perm Descriptor attribute access permissions,
+ *		 a bitmap of enum bt_gatt_perm values.
+ *  @param _read Descriptor attribute read callback
+ *               (bt_gatt_attr_read_func_t).
+ *  @param _write Descriptor attribute write callback
+ *                (bt_gatt_attr_write_func_t).
```
Is it possible to refer to a type using doxygen? Perhaps just `@ref bt_gatt_attr_read_func_t`, or something?

alwa-nordic replied:
```suggestion
 *               (@ref bt_gatt_attr_read_func_t).
 *  @param _write Descriptor attribute write callback
 *                (@ref bt_gatt_attr_write_func_t).
```

alwa-nordic commented on include/bluetooth/gatt.h:
```diff
@@ -882,7 +899,8 @@ ssize_t bt_gatt_attr_read_cud(struct bt_conn *conn,
  *  @param _value User description NULL-terminated C string.
- *  @param _perm Descriptor attribute access permissions.
+ *  @param _perm Descriptor attribute access permissions,
+ *		 a bitmap of enum bt_gatt_perm values.
```
```suggestion
 *		 a bitmap of @ref bt_gatt_perm values.
```

alwa-nordic commented on include/bluetooth/gatt.h:
```diff
@@ -939,9 +960,10 @@ ssize_t bt_gatt_attr_read_cpf(struct bt_conn *conn,
- *  @param _write Attribute write callback.
+ *  @param _perm Attribute access permissions,
+ *		 a bitmap of enum bt_gatt_perm values.
+ *  @param _read Attribute read callback (bt_gatt_attr_read_func_t).
```
```suggestion
 *  @param _read Attribute read callback (@ref bt_gatt_attr_read_func_t).
```

alwa-nordic commented on include/bluetooth/gatt.h:
```diff
@@ -939,9 +960,10 @@ ssize_t bt_gatt_attr_read_cpf(struct bt_conn *conn,
+ *  @param _perm Attribute access permissions,
+ *		 a bitmap of enum bt_gatt_perm values.
+ *  @param _read Attribute read callback (bt_gatt_attr_read_func_t).
+ *  @param _write Attribute write callback (bt_gatt_attr_write_func_t).
```
```suggestion
 *  @param _write Attribute write callback (@ref bt_gatt_attr_write_func_t).
```

alwa-nordic commented on include/bluetooth/gatt.h:
```diff
@@ -818,7 +834,8 @@ ssize_t bt_gatt_attr_write_ccc(struct bt_conn *conn,
  *  @param _changed Configuration changed callback.
- *  @param _perm CCC access permissions.
+ *  @param _perm CCC access permissions,
+ *		 a bitmap of enum bt_gatt_perm values.
```
```suggestion
 *		 a bitmap of @ref bt_gatt_perm values.
```

alwa-nordic commented on include/bluetooth/gatt.h:
```diff
@@ -805,7 +820,8 @@ ssize_t bt_gatt_attr_write_ccc(struct bt_conn *conn,
  *  @param _ccc CCC attribute user data, shall point to a _bt_gatt_ccc.
- *  @param _perm CCC access permissions.
+ *  @param _perm CCC access permissions,
+ *		 a bitmap of enum bt_gatt_perm values.
```
```suggestion
 *		 a bitmap of @ref bt_gatt_perm values.
```

alwa-nordic commented on include/bluetooth/gatt.h:
```diff
@@ -667,10 +678,14 @@ ssize_t bt_gatt_attr_read_chrc(struct bt_conn *conn,
+ *  @param _perm Characteristic Attribute access permissions,
+ *		 a bitmap of enum bt_gatt_perm values.
+ *  @param _read Characteristic Attribute read callback
+ *               (bt_gatt_attr_read_func_t).
```
```suggestion
 *               (@ref bt_gatt_attr_read_func_t).
```

alwa-nordic commented on include/bluetooth/gatt.h:
```diff
@@ -667,10 +678,14 @@ ssize_t bt_gatt_attr_read_chrc(struct bt_conn *conn,
+ *  @param _read Characteristic Attribute read callback
+ *               (bt_gatt_attr_read_func_t).
+ *  @param _write Characteristic Attribute write callback
+ *                (bt_gatt_attr_write_func_t).
```
```suggestion
 *                (@ref bt_gatt_attr_write_func_t).
```

spbnick committed:
doc: Fix a couple typos

Fix a couple small typos in docs.

Signed-off-by: Nikolai Kondrashov <spbnick@gmail.com>

Thalley commented:
@spbnick Planning on updating this PR? Otherwise someone else should take over, as the additions are definitely welcome

spbnick commented:
@Thalley, I'll try to find the time over the coming weekend, but if I don't I'd suppose this should be free for picking. Thanks!

Thalley commented:
> @Thalley, I'll try to find the time over the coming weekend, but if I don't I'd suppose this should be free for picking. Thanks!

@spbnick Any updates?

spbnick committed:
Bluetooth: GATT: Document some macro arguments

Name some types in bluetooth/gatt.h and use those names to document the
expected values of some of the BT_GATT_* macro arguments, in an attempt
to make understanding the API a little easier.

Signed-off-by: Nikolai Kondrashov <spbnick@gmail.com>

spbnick committed:
Bluetooth: GATT: Refer to BT_GATT_WRITE_FLAG_*

Refer to BT_GATT_WRITE_FLAG_* specifically, instead of the more general
BT_GATT_WRITE_* namespace to avoid future clashes with new symbols
there.

Signed-off-by: Nikolai Kondrashov <spbnick@gmail.com>

spbnick commented:
Thanks for the reminder, @github-actions, @Thalley. I took a break from my HW work and tried to get this in shape. Let's see what CI says :)

spbnick commented:
Please take a look again, whoever is available, before this goes stale once more :grin:

Thalley commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -666,10 +676,14 @@ ssize_t bt_gatt_attr_read_chrc(struct bt_conn *conn,
- *  @param _perm Characteristic Attribute access permissions.
- *  @param _read Characteristic Attribute read callback.
- *  @param _write Characteristic Attribute write callback.
+ *  @param _props Characteristic attribute properties,
+ *		  a bitmap of @c BT_GATT_CHRC_* macros.
+ *  @param _perm Characteristic Attribute access permissions,
+ *		 a bitmap of @ref bt_gatt_perm values.
```
```suggestion
 *  @param _props Characteristic attribute properties,
 *                a bitmap of @c BT_GATT_CHRC_* macros.
 *  @param _perm Characteristic Attribute access permissions,
 *               a bitmap of @ref bt_gatt_perm values.
```

Please use spaces instead of tabs when indenting after a non-whitespace character

spbnick replied:
Sure. Fixed all of these, I think.

However, what's the actual full rule about tabs after non-whitespace characters? Because there's plenty of those in that very file beside my changes: aligning macro definitions, field names in structs and newline-escaping backslashes. Is this only about doxygen comments?

Out of all things Zephyr took from the Linux kernel, I wish tabs weren't one of them :joy:

Thalley commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -804,7 +818,8 @@ ssize_t bt_gatt_attr_write_ccc(struct bt_conn *conn,
  *  @param _ccc CCC attribute user data, shall point to a _bt_gatt_ccc.
- *  @param _perm CCC access permissions.
+ *  @param _perm CCC access permissions,
+ *		 a bitmap of @ref bt_gatt_perm values.
```
Indent with spaces instead of tabs

Thalley replied:
Looks like this applies a few more places

Thalley reviewed (changes requested):
A few minor indentation fixes, otherwise LGTM

alwa-nordic commented:
> > Can you try `@c BT_GATT_CHRC_*` (make it a literal)?
> 
> [This helped](https://github.com/zephyrproject-rtos/zephyr/pull/40436/commits/4fe3074c62129c754aad6ff7e89de336981f42f5) (docs built), and I'll use it, but I would still like to know what was going on joy.

We have markdown enabled in doxygen. `*some words*` is synax for emphasis. Maybe it barfed on the unbalanced `*`. (I haven't tested this theory.)

spbnick commented:
> > > Can you try `@c BT_GATT_CHRC_*` (make it a literal)?
> > 
> > 
> > [This helped](https://github.com/zephyrproject-rtos/zephyr/pull/40436/commits/4fe3074c62129c754aad6ff7e89de336981f42f5) (docs built), and I'll use it, but I would still like to know what was going on joy.
> 
> We have markdown enabled in doxygen. `*some words*` is synax for emphasis. Maybe it barfed on the unbalanced `*`. (I haven't tested this theory.)

For better or worse, it doesn't reproduce anymore :man_shrugging: So I rolled back that `@c`.

spbnick commented:
Thank you for the reviews and comments everyone! Hoping this will get merged now :grin::crossed_fingers:
